pkg/new/onit/env/network.go

```diff
@@ -41,18 +63,246 @@ type network struct {
 
 // networkSetup is an implementation of the NetworkSetup interface
 type networkSetup struct {
 	*serviceSetup
-	name string
+	name     string
+	topoType TopoType
+	nodes    int
+}
+
+func (s *networkSetup) Single() setup.NetworkSetup {
+	s.topoType = Single
+	return s
+}
+
+func (s *networkSetup) Linear(nodes int) setup.NetworkSetup {
+	s.topoType = Single
+	s.nodes = nodes
+	return s
 }
 
 func (s *networkSetup) Using() setup.ServiceSetup {
 	return s
 }
 
 func (s *networkSetup) Setup() error {
+	if err := s.createPod(); err != nil {
+		return err
+	}
+	if err := s.createService(); err != nil {
+		return err
+	}
+	if err := s.awaitReady(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createPod creates a stratum network pod
+func (s *networkSetup) createPod() error {
+	var topoSpec string
+	switch s.topoType {
+	case Single:
+		topoSpec = "single"
+	case Linear:
+		topoSpec = fmt.Sprintf("linear,%d", s.nodes)
+	}
+
+	var isPrivileged = true
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      s.name,
+			Namespace: s.namespace,
+			Labels: map[string]string{
+				"type":    "network",
+				"network": s.name,
+			},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:            "stratum-simulator",
+					Image:           s.image,
+					ImagePullPolicy: s.pullPolicy,
+					Stdin:           true,
+					Args:            []string{"--topo", topoSpec},
+					Ports: []corev1.ContainerPort{
+						{
+							Name:          "stratum",
+							ContainerPort: 50001,
+						},
+					},
+					ReadinessProbe: &corev1.Probe{
+						Handler: corev1.Handler{
+							TCPSocket: &corev1.TCPSocketAction{
+								Port: intstr.FromInt(50001),
+							},
+						},
+						InitialDelaySeconds: 5,
+						PeriodSeconds:       10,
+					},
+					LivenessProbe: &corev1.Probe{
+						Handler: corev1.Handler{
+							TCPSocket: &corev1.TCPSocketAction{
+								Port: intstr.FromInt(50001),
+							},
+						},
+						InitialDelaySeconds: 15,
+						PeriodSeconds:       20,
+					},
+					SecurityContext: &corev1.SecurityContext{
+						Privileged: &isPrivileged,
+					},
+				},
+			},
+		},
+	}
+
+	_, err := s.kubeClient.CoreV1().Pods(s.namespace).Create(pod)
+	return err
+}
+
+// awaitReady waits for the given simulator to complete startup
+func (s *networkSetup) awaitReady() error {
+	for {
+		pod, err := s.kubeClient.CoreV1().Pods(s.namespace).Get(s.name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		} else if len(pod.Status.ContainerStatuses) > 0 && pod.Status.ContainerStatuses[0].Ready {
+			return nil
+		} else {
+			time.Sleep(100 * time.Millisecond)
+		}
+	}
+}
+
+// getNumDevices returns the number of devices in the topology
+func (s *networkSetup) getNumDevices() int {
+	switch s.topoType {
+	case Single:
+		return 1
+	case Linear:
+		return s.nodes
+	}
+	return 0
+}
+
+// createService creates a network service
+func (s *networkSetup) createService() error {
+	var port int32 = 50001
```

Comment:
Should we include this port in networkSetup structure?

Reply:
Hmm... I think the network setup thing needs to be revisited. Will have to look at the devices/services again when I get there. Going to merge this and then start going through all the features.